src/twisted/trial/_dist/test/test_worker.py

```diff
@@ -86,197 +76,155 @@ class LocalWorkerAMPTests(TestCase):
+            result,
+            matches_result(
+                skips=contains_exactly(  # type: ignore[arg-type]
+                    isTuple(  # type: ignore[arg-type]
+                        equal_to(case),
+                        equal_to("pyunit skip"),
+                    ),
+                ),
+            ),
         )
-        d.addCallback(lambda result: results.append(result["success"]))
-        self.pumpTransports()
-
-        self.assertEqual(self.testCase, self.result.skips[0][0])
-        self.assertTrue(results)
 
-    def test_runUnexpectedSuccesses(self):
+    def test_runUnexpectedSuccesses(self) -> None:
         """
         Run a test, and succeed unexpectedly.
         """
-        results = []
-
-        d = self.worker.callRemote(
-            managercommands.AddUnexpectedSuccess, testName=self.testName, todo="todo"
+        case = skipping.SynchronousStrictTodo("test_todo7")
+        result = self.workerRunTest(case)
+        assert_that(
+            result,
+            matches_result(
+                unexpectedSuccesses=contains_exactly(  # type: ignore[arg-type]
+                    isTuple(  # type: ignore[arg-type]
+                        equal_to(case),
+                        equal_to("todo7"),
+                    ),
+                ),
+            ),
         )
-        d.addCallback(lambda result: results.append(result["success"]))
-        self.pumpTransports()
-
-        self.assertEqual(self.testCase, self.result.unexpectedSuccesses[0][0])
-        self.assertTrue(results)
 
-    def test_testWrite(self):
+    def test_testWrite(self) -> None:
         """
         L{LocalWorkerAMP.testWrite} writes the data received to its test
         stream.
         """
-        results = []
         stream = StringIO()
         self.managerAMP.setTestStream(stream)
-
-        command = managercommands.TestWrite
-        d = self.worker.callRemote(command, out="Some output")
-        d.addCallback(lambda result: results.append(result["success"]))
-        self.pumpTransports()
-
+        d = self.worker.callRemote(managercommands.TestWrite, out="Some output")
+        self.flush()
+        self.assertEqual({"success": True}, self.successResultOf(d))
         self.assertEqual("Some output\n", stream.getvalue())
-        self.assertTrue(results)
 
-    def test_stopAfterRun(self):
+    def test_stopAfterRun(self) -> None:
         """
         L{LocalWorkerAMP.run} calls C{stopTest} on its test result once the
         C{Run} commands has succeeded.
         """
-        result = object()
-        stopped = []
-
-        def fakeCallRemote(command, testCase):
-            return succeed(result)
-
-        self.managerAMP.callRemote = fakeCallRemote
 
         class StopTestResult(TestResult):
-            def stopTest(self, test):
-                stopped.append(test)
+            _stopped = False
+
+            def stopTest(self, test: PyUnitTestCase) -> None:
+                self._stopped = True
```

Comment:
```suggestion
            _stopped = []

            def stopTest(self, test: PyUnitTestCase) -> None:
                self._stopped.append(True)
```

I am still using append in tests at it can help detect corner cases in which something is called multiple times.

The append might look strange on a class variable, but this class should only be used once, inside this test.

The test code looks uglier ... but I think is best to have as strict of an assertion as possible.

The test code should be easy to maintain, but at the same time is there to prevent unexpected behaviour.

Reply:
Perhaps an integer counter?

Reply:
I put the list back because the original test also asserted the right test case was passed as an argument.